### PR TITLE
Add testing of sal.dataclass

### LINKED ISSUE #1: Use setuptools entry points to handle dataclasses
SAL core could potentially dynamically load data classes from installed packages by registering a 'sal.dataclass' entry point in each data class defining package:

https://setuptools.readthedocs.io/en/latest/setuptools.html#dynamic-discovery-of-services-and-plugins

### LINKED ISSUE #72: Add testing of SAL dataclass
### Topic

Tests of the SAL dataclass submodules

### Description of Tests

Tests of:
- Class attribute setting i.e. do all `DataClass` subclasses implement `CLASS`, `GROUP`, `VERSION`, `TYPE`
- `to_dict` and `from_dict` implementations
- `summary` implementations produce the correct `SummaryObject`
- Behaviour specific to each `DataClass` subclass.  This is commonly relating to parameter validation.

### Type

Unit

### PR DESCRIPTION
**Pull request by TOFarmer**
### Description of work

Adds testing of dataclass submodule.

Tests were added for all `DataClass` subclasses, except for `Dictionary` and `Array`, for which tests already exist.  At a later point these tests should be converted to the new testing approach.

The `xfail` behaviour in both `test_array_dimension_unsupported_dtype` and `test_array_dimension_multidimensional` is covered by #74 

### Fixes

Fixes #72 

### To test

- [ ] Check that subclass specific behaviour is adequately tested. 

### Reviewer checklist

- [ ] The author suggested tests are successful.
- [ ] Relevant documentation (user and developer) has been added, and is clear and complete.
- [ ] Suitable unit tests have been added.  If unit tests are not included in this MR, please comment as to why and create an issue to address this.

### CODE REVIEW

File: sal/tests/dataclass/unit/test_calculated_dimension.py [hunk 1]
```diff
@@ -0,0 +1,14 @@
+import numpy as np
+import pytest
+
+from sal.dataclass import CalculatedDimension
+
+
+@pytest.mark.parametrize('length, start, step, expected',
+                         [(5, 1, 2, np.array([1., 3., 5., 7., 9.])),
+                          (3, 10, 10, np.array([10., 20., 30.])),
+                          (4, 6, -2, np.array([6., 4., 2., 0.]))])
```
Comment (thomassclayton) [inline sal/tests/dataclass/unit/test_calculated_dimension.py:10]: **suggestion:** Just to test edge cases, you could add runs with lengths of 0 and 1.

File: sal/tests/dataclass/unit/test_common.py [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+     ('constant_error_summary', 'error_constant', 'signal_error', 1),
+     ('sym_array_error', 'error_symmetrical', 'signal_error', 1),
+     ('sym_array_error_summary', 'error_symmetrical', 'signal_error', 1),
+     ('array_dim', 'coordinate_array', 'signal_dimension', 1),
+     ('array_dim_summary', 'coordinate_array', 'signal_dimension', 1),
+     ('calc_dim', 'coordinate_calc', 'signal_dimension', 1),
+     ('calc_dim_summary', 'coordinate_calc', 'signal_dimension', 1),
+     ('signal', 'signal', 'signal', 1),
+     ('signal', 'signal', 'signal', 1)])
+def test_class_attributes(fixture, class_, group, version, request):
```
Comment (thomassclayton) [inline sal/tests/dataclass/unit/test_common.py:30]: **question:** `test_array.py` tests setting the description, was it decided that this was not necessary for the new tests?

**Review by thomassclayton — COMMENTED** [review]
There are `# TODO: add tests` comments all over the dataclasses, which can now be removed for those that have been tested.

Aside from that and the other comments, looks fine to me.

File: sal/tests/dataclass/unit/test_common.py (hunk 2 above)
Comment (TOFarmer) [reply sal/tests/dataclass/unit/test_common.py:30]: Yeah, it is either a simple assignment or a redundant conditional, which has been removed in another branch I'm working on.  Whether or not `description` is set correctly is also obliquely tested in the `to_dict` and `from_dict` tests, so it is covered anyway, albeit not in the way it would be if I thought it was necessary to test.

File: sal/tests/dataclass/unit/test_calculated_dimension.py (hunk 1 above)
Comment (TOFarmer) [reply sal/tests/dataclass/unit/test_calculated_dimension.py]: Good point.  As length of 0 is not allowed I've created a separate test for this.

**Comment by TOFarmer** [comment]
> There are `# TODO: add tests` comments all over the dataclasses, which can now be removed for those that have been tested.

Ah yes, thanks, I do like removing lingering `#TODO` statements!

**Comment by TOFarmer** [comment]
Although actually, I think I'll remove those `# TODO` statements in the `refactor_node_elements` branch, just so I don't have any conflicts.
